crates/gen/src/types/ntstatus.rs

```diff
@@ -0,0 +1,34 @@
+use super::*;
+
+pub fn gen_ntstatus() -> TokenStream {
+    quote! {
+        #[repr(transparent)]
+        #[derive(::std::default::Default, ::std::clone::Clone, ::std::marker::Copy, ::std::cmp::PartialEq, ::std::cmp::Eq, ::std::fmt::Debug)]
+        pub struct NTSTATUS(pub u32);
+
+        impl NTSTATUS {
+            #[inline]
+            pub const fn is_ok(self) -> bool {
+                self.0 == 0
+            }
+
+            #[inline]
+            pub const fn is_err(self) -> bool {
+                !self.is_ok()
+            }
+
+            #[inline]
+            pub fn ok(self) -> ::windows::Result<()> {
+                if self.is_ok() {
+                    Ok(())
```

Comment:
How about changing this to return `::windows::Result<NTSTATUS>` and `Ok(self)`? The benefits are that clients can continue to use the `?` operator for error propagation in case of true errors, but still get access to a non-zero success code.

Reply:
I wonder how common this is? How often would this end up being used? Is there a representative example?

Reply:
I'm not aware of one. There are about 80 non-zero status codes with severity 0 (success) and around 100 with severity 1 (information) defined in *ntstatus.h*. While I did recognize some (e.g. `STATUS_PENDING`) I wouldn't know of a Windows API call where those actually show up as the *return value*.

Things are different with driver development, though, where functions like [IoRegisterDeviceInterface](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-ioregisterdeviceinterface) can indeed return values like `STATUS_OBJECT_NAME_EXISTS` (0x40000000). I don't know whether driver development is in scope for the Windows crate.

Reply:
Hmm, perhaps it would be better to keep the `.ok()` method but remove the automatic `Result` transformation so that developers can easily call `method().ok()` for most cases but deal directly with the error code in other cases. 

`NTSTATUS` is much the same as `HRESULT` in its structure and usage but it certainly does seem to have a lot more use for success codes. I tried writing a simple wrapper for `BCryptVerifySignature` but its certainly not convenient. 

We'd like to eventually support driver development

Reply:
Perhaps just a `to_hresult` method on NTSTATUS to make this a littler simpler:

```rust
fn verify() -> Result<bool> {
    match BCryptVerifySignature() {
        Err(e) => {
            if e.code() == HRESULT(STATUS_INVALID_SIGNATURE.0 | 0x1000_0000) {
                Ok(false)
            } else {
                Err(e)
            }
        }
        _ => Ok(true),
    }
}
```

Reply:
And ironically the docs state that it could return `NTE_NO_MEMORY`, which is actually an `HRESULT`. 😉

Reply:
https://github.com/microsoft/windows-rs/pull/997